### PR TITLE
 Refine distributed lock to lock only engine bootstrap

### PR DESCRIPTION
**Pull request by yaooqinn**
### _Why are the changes needed?_

1. refine the zk lock to lock only during engine bootstrapping, not wrap the part that only gets the address of engines. This can significantly reduce overhead.

2. add tests for the locking proccess.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+ *
+ * @param conf Engine configuration
+ * @param user Caller of the engine
+ * @param sessionId Id of the corresponding session in which the engine is created
+ */
+private[kyuubi] class Engine private(conf: KyuubiConf, user: String, sessionId: String)
+  extends Logging {
+
+  // The corresponding ServerSpace where the engine belongs to
+  private val serverSpace: String = conf.get(HA_ZK_NAMESPACE)
```
Comment (pan3793) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala:50]: personally, i prefer name it `serverNamespace`

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    case _ => subDomain match {
+      case Some(domain) => ZKPaths.makePath(s"${serverSpace}_$shareLevel", appUser, domain)
+      case None => ZKPaths.makePath(s"${serverSpace}_$shareLevel", appUser)
+    }
+  }
+
+  /**
+   * The distributed lock path used to ensure only once engine being created for non-CONNECTION
+   * share level.
+   *   /$serverSpace_USER/lock/$user(/$subDomain)?
```
Comment (pan3793) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala:98]: `/${serverSpace}_USER/lock/${user}[/${subDomain}]`

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+import org.apache.kyuubi.session.SessionHandle
+
+/**
+ * The description and functionality of an engine at server side
+ *
+ * @param conf Engine configuration
+ * @param user Caller of the engine
+ * @param sessionId Id of the corresponding session in which the engine is created
+ */
+private[kyuubi] class Engine private(conf: KyuubiConf, user: String, sessionId: String)
```
Comment (pan3793) [inline kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala:46]: > The description and functionality of an engine at server side

How about name it `EngineRef`?

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        }
+      }
+  }
+
+  def getEngineAddr(zkClient: CuratorFramework): Option[(String, Int)] = {
+    getServerHost(zkClient, engineSpace)
+  }
+
+  def start(zkClient: CuratorFramework): (String, Int) = tryWithLock(zkClient) {
+    var sh = getEngineAddr(zkClient)
```
[the remark this answers is not in the document]
Comment (Honglun) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala:133]: server host?

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala (hunk 1 above)
Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala]: we've create the terminology

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala (hunk 3 above)
Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala]: sgtm

File: kyuubi-main/src/test/scala/org/apache/kyuubi/engine/EngineSuite.scala [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        }
+      }
+    }
+    val factory = new NamedThreadFactory("engine-test", false)
+    val thread1 = factory.newThread(r1)
+    val thread2 = factory.newThread(r2)
+    thread1.start()
+    thread2.start()
+
+    eventually(timeout(90.seconds), interval(100.microseconds)) {
```
[the remark this answers is not in the document]
Comment (yaooqinn) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/engine/EngineSuite.scala]: 1000 millis is 1 s

File: kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (yaooqinn) [reply kyuubi-main/src/main/scala/org/apache/kyuubi/engine/Engine.scala]: renamed to engineRef

File: kyuubi-main/src/test/scala/org/apache/kyuubi/engine/EngineSuite.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (pan3793) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/engine/EngineSuite.scala]: oops... i see

**Review by pan3793 — APPROVED** [review]

**Review by Honglun — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged for v1.2.0

**Pull request closed without merging**